File: plugins/org.eclipse.tm.tcf.core/src/org/eclipse/tm/internal/tcf/services/remote/LineNumbersProxy.java
package org.eclipse.tm.internal.tcf.services.remote;

import java.util.Collection;
import java.util.Map;

import org.eclipse.tm.tcf.core.Command;
import org.eclipse.tm.tcf.protocol.IChannel;
import org.eclipse.tm.tcf.protocol.IToken;
import org.eclipse.tm.tcf.services.ILineNumbers;


public class LineNumbersProxy implements ILineNumbers {
    
    private final IChannel channel;

    public LineNumbersProxy(IChannel channel) {
        this.channel = channel;
    }

    public String getName() {
        return NAME;
    }

    public IToken mapToSource(String context_id, Number start_address,
            Number end_address, final DoneMapToSource done) {
        return new Command(channel, this, "mapToSource", new Object[]{ context_id,
                start_address, end_address }) {
            @Override
            public void done(Exception error, Object[] args) {
                CodeArea[] arr = null;
                if (error == null) {
                    assert args.length == 2;
                    error = toError(args[0]);
                    arr = toTextAreaArray(args[1]);
                }
                done.doneMapToSource(token, error, arr);
            }
        }.token;
    }
    
    private static int getInteger(Map<String,Object> map, String name, int def) {
        Number n = (Number)map.get(name);
        if (n == null) return def;
        return n.intValue();
    }
    
    private static String getString(Map<String,Object> map, String name, String def) {
        String s = (String)map.get(name);
        if (s == null) return def;
        return s;
    }
    
    private static boolean getBoolean(Map<String,Object> map, String name) {
        Boolean b = (Boolean)map.get(name);
        if (b == null) return false;
        return b.booleanValue();
    }
    
    @SuppressWarnings("unchecked")
    private CodeArea[] toTextAreaArray(Object o) {
        if (o == null) return null;
        Collection<Map<String,Object>> c = (Collection<Map<String,Object>>)o;
        int n = 0;
        CodeArea[] arr = new CodeArea[c.size()];
        String directory = null;
        String file = null;
        for (Map<String,Object> area : c) {
            directory = getString(area, "Dir", directory);
            file = getString(area, "File", file);
            arr[n++] = new CodeArea(directory, file,
                    getInteger(area, "SLine", 0), getInteger(area, "SCol", 0),
                    getInteger(area, "ELine", 0), getInteger(area, "ECol", 0),
                    (Number)area.get("SAddr"), (Number)area.get("EAddr"),
                    getInteger(area, "ISA", 0),
                    getBoolean(area, "IsStmt"), getBoolean(area, "BasicBlock"),
                    getBoolean(area, "PrologueEnd"), getBoolean(area, "EpilogueBegin"));
        }
        return arr;
    }
}

File: python/src/tcf/services/remote/LineNumbersProxy.py
# *******************************************************************************
# * Copyright (c) 2011 Wind River Systems, Inc. and others.
# * All rights reserved. This program and the accompanying materials
# * are made available under the terms of the Eclipse Public License v1.0
# * which accompanies this distribution, and is available at
# * http://www.eclipse.org/legal/epl-v10.html
# *
# * Contributors:
# *     Wind River Systems - initial API and implementation
# *******************************************************************************

from tcf.services import linenumbers
from tcf.channel.Command import Command

class LineNumbersProxy(linenumbers.LineNumbersService):

    def __init__(self, channel):
        self.channel = channel

    def mapToSource(self,  context_id, start_address, end_address, done):
        done = self._makeCallback(done)
        service = self
        class MapCommand(Command):
            def __init__(self):
                super(MapCommand, self).__init__(service.channel, service, 
                        "mapToSource", (context_id, start_address, end_address))
            def done(self, error, args):
                arr = None
                if not error:
                    assert len(args) == 2
                    error = self.toError(args[0])
                    arr = _toCodeAreaArray(args[1])
                done.doneMapToSource(self.token, error, arr)
        return MapCommand().token

    def mapToMemory(self, context_id, file, line, column, done):
        done = self._makeCallback(done)
        service = self
        class MapCommand(Command):
            def __init__(self):
                super(MapCommand, self).__init__(service.channel, service, 
                        "mapToMemory", (context_id, file, line, column))
            def done(self, error, args):
                arr = None
                if not error:
                    assert len(args) == 2
                    error = self.toError(args[0])
                    arr = _toCodeAreaArray(args[1])
                done.doneMapToMemory(self.token, error, arr)
        return MapCommand().token

def _toCodeAreaArray(o):
    if not o: return None
    arr = []
    directory = None
    file = None
    for area in o:
        directory = area.get("Dir", directory)
        file = area.get("File", file)
        arr.append(linenumbers.CodeArea(directory, file,
                area.get("SLine", 0), area.get("SCol", 0),
                area.get("ELine", 0), area.get("ECol", 0),
                area.get("SAddr"), area.get("EAddr"),
                area.get("ISA", 0),
                area.get("IsStmt"), area.get("BasicBlock"),
                area.get("PrologueEnd"), area.get("EpilogueBegin")))
    return arr
